File: server/src/main/java/server/xodus/MessageDatabase.java
package server.xodus;

import jetbrains.exodus.ArrayByteIterable;
import jetbrains.exodus.env.Cursor;
import jetbrains.exodus.env.Store;
import lib.SignedMessage;
import lib.message.EnvelopeMessage;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class MessageDatabase {

    private final Store store;

    public MessageDatabase(Store store) {
        this.store = store;
    }

    public void addMessage(byte[] target, byte[] message) {
        store.getEnvironment().executeInTransaction(txn -> store.put(txn,
                new ArrayByteIterable(target), new ArrayByteIterable(message)));
    }

    public List<SignedMessage<EnvelopeMessage>> getMessages(byte[] target) {
        return store.getEnvironment().computeInReadonlyTransaction(txn -> {
            var list = new ArrayList<ArrayByteIterable>();

            Cursor cursor = store.openCursor(txn);
            var m = cursor.getSearchKey(new ArrayByteIterable(target));
            if (m != null) {
                list.add(new ArrayByteIterable(m));
//                cursor.deleteCurrent();

                while (cursor.getNextDup()) {
                    list.add(new ArrayByteIterable(cursor.getValue()));
//                    cursor.deleteCurrent();
                }
            }

            return list.stream().map(o -> new SignedMessage<EnvelopeMessage>(o.getBytesUnsafe()))
                    .collect(Collectors.toList());
        });
    }
}

File: server/src/main/kotlin/server/xodus/MessageDatabase.kt
package server.xodus

import jetbrains.exodus.ArrayByteIterable
import jetbrains.exodus.bindings.StringBinding
import jetbrains.exodus.env.Store
import jetbrains.exodus.env.Transaction
import kotlinx.serialization.json.Json
import lib.SendMessageEnvelope
import lib.SignedMessage
import java.util.stream.Collectors

class MessageDatabase(private val store: Store) {
    fun addMessage(target: ByteArray, message: String) {
        store.environment.executeInTransaction { txn ->
            store.put(
                txn,
                ArrayByteIterable(target), StringBinding.stringToEntry(message)
            )
        }
    }

    fun getMessages(target: ByteArray): List<SignedMessage<SendMessageEnvelope>> {
        return store.environment.computeInReadonlyTransaction { txn: Transaction? ->
            val list: ArrayList<ArrayByteIterable> = ArrayList<ArrayByteIterable>()
            val cursor = store.openCursor(txn!!)
            val m = cursor.getSearchKey(ArrayByteIterable(target))
            if (m != null) {
                list.add(ArrayByteIterable(m))

                //                cursor.deleteCurrent();
                while (cursor.nextDup) {
                    list.add(ArrayByteIterable(cursor.value))
                    //                    cursor.deleteCurrent();
                }
            }
            list.stream()
                .map {
                    Json.decodeFromString<SignedMessage<SendMessageEnvelope>>(
                        StringBinding.entryToString(it)
                    )
                }
                .collect(Collectors.toList())
        }
    }
}
